---
title: "Figure 4"
author: '`r Sys.info()["user"]`'
date: '`r format(Sys.time(), "%B %d, %Y %H:%M:%S %Z")`'
output:
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
    highlight: pygments
    df_print: kable
    keep_md: true
params:
  path_to_data: ~
  accessibility: "lock"
  path_to_TCR: ~
  McPAS: ~

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE,
	dev = c("png")
)
```


# Introduction

## Load libraries


```{r lib}
library(Seurat)
library(canceRbits)
library(dplyr)
library(patchwork)
library(DT)
library(SCpubr)
library(tibble)
library(dittoSeq)
library(scRepertoire)
library(reshape2)
library(viridis)
library(tidyr)

library(grDevices)
library(ggpubr)
library(ggplot2)
library(ggnewscale)

library(RColorBrewer)
library(scales)

library(enrichplot)
library(clusterProfiler)
library("org.Hs.eg.db")
library(DOSE)
library(msigdbr)
library(stringr)
```

## Define colors


```{r  COLORS, echo=FALSE, fig.height=8, fig.width=8, out.width='100%'}
colors_anno_l2 <- c(
   "Tregs"                = "#ae017e",
   "other"                = "#d9d9d9",
   "Memory T cells"       =  "#377eb8",
   "Cytotoxic T cells"    =  "#f781bf",  
   "T helper cells"       = "#fed976",
   "Naive CD4+ T cells"   = "#a6cee3",
   "Proliferating cells"  = "#d0d1e6",
   "Exhausted T cells"    = "#e31a1c",
   "Natural killer cells" = "#9e9ac8",
   "gd T cells"           = "#1d91c0")

#"TH-like CD4+ T cells" = "#bdbdbd"

colors_clonotype = c("None ( < X <= 0)"    = "white",
                     "Rare (0 < X <= 5e-05)" = "grey",
                     "Small (5e-05 < X <= 5e-04)"   = "#8c6bb1",
                     "Medium (5e-04 < X <= 0.005)"   = "#41b6c4",
                     "Large (0.005 < X <= 0.05)"      = "#fec44f",
                     "Hyperexpanded (0.05 < X <= 1)" = "#ce1256"
)

colors_McPAS = c("hHSV: CAVGSARQLTF_CSVWAGGTQPQHF" = "#35978f",
           "hHSV: CAVNPDSWGKLQF_CATSVRDEQYF" = "#66bd63",
           "Cancer" = "#fdae61"
)

```


## Load single cell RNA-Seq data

```{r}
# Load seurat object containing single cell pre-processed and annotated data, in the metadata folder

srat <- readRDS(params$path_to_data)

if(params$accessibility == "unlock"){
  combined <- readRDS(params$path_to_TCR)
}

meta <- srat@meta.data
meta$WHO <- "SD"
meta$WHO[meta$patient %in% c("NeoBCC007_post", "NeoBCC008_post", "NeoBCC012_post", "NeoBCC017_post")] <- "CR"
meta$WHO[meta$patient %in% c("NeoBCC004_post", "NeoBCC006_post", "NeoBCC010_post", "NeoBCC011_post")] <- "PR"
srat <- AddMetaData(srat, meta$WHO, col.name = "WHO")
srat$WHO <- factor(srat$WHO, levels = c("CR", "PR", "SD"))
```



# Figure 4a

```{r  Figure4a, echo=FALSE, fig.height=8, fig.width=8, out.width='50%'}
s    <- subset(srat, subset = anno_l1 %in% c("CD8+ T cells","CD4+ T cells", "Proliferating cells", "Tregs", "Natural killer cells"))

# define level order
s@meta.data$anno_l2 <- factor(s@meta.data$anno_l2, 
                              levels=c("Naive CD4+ T cells",
                                       "T helper cells", 
                                       "Memory T cells",
                                       "Tregs",         
                                       "Exhausted T cells",
                                       "Cytotoxic T cells",
                                       "Natural killer cells" ,
                                       "Proliferating cells",
                                       "gd T cells"
                                       ))

# define order of plotting T cell subsets
order = c("Tregs", 
          "gd T cells",   
          "Naive CD4+ T cells", 
          "Exhausted T cells",
          "Natural killer cells",
          "Cytotoxic T cells",
          "Memory T cells",  
          "T helper cells",
          "Proliferating cells")


p <- SCpubr::do_DimPlot(sample = s, 
                   reduction = "umap", 
                   label = TRUE, 
                   repel = TRUE, 
                   label.color = "black", 
                   legend.position = "none", 
                   group.by = "anno_l2", 
                   font.size = 30, 
                   pt.size = 0.5, 
                   colors.use = colors_anno_l2,
                   shuffle = FALSE, 
                   order = order) + theme_minimal()  + NoLegend() + theme(text = element_text(size=30))
p

DT::datatable(p$data, 
              caption = ("Figure 4a"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))

```



# Figure 4b

```{r Figure4b, echo=FALSE, fig.height=10, fig.width=26, message=FALSE, warning=FALSE, out.width='100%'}

gene_list <- list( "Naive" = c("LEF1", "SELL"),
                   "T helper" = c("CD3E", "CD4", "IL7R"),
                   "TH1" = c("IFNG", "TBX21"),
                   "TH2" = c("GATA3", "IL13"),
                   "TH17" =  c("KLRB1", "IL17A"),
                   "TFH"= c("TCF7", "BCL6",  "CXCR5", "LEF1"),
                   "M" = c("CD44"),
                   "Treg" = c("FOXP3", "IL2RA","TNFRSF4", "TNFRSF18"),
                   "Exhausted T" = c("PDCD1", "CTLA4", "LAG3", "TIGIT", "HAVCR2", "TIM3", "CD39", "ENTPD1", "ITGAE", "ICOS"),
                   "Cytotoxic" = c("CD8A", "GZMA", "GZMB","GZMK"),
                   "NK"=c("PRF1", "GNLY"),
                   "gdT" = c("TRDC"),
                   "P" = c("MKI67")
                  )

s@meta.data$anno_l3 <- factor(s@meta.data$anno_l3, 
                              levels=c("Naive CD4+ T cells",
                                       "TH-like CD4+ T cells",
                                       "TH1","TH2", "TH17", "TFH",
                                       "Memory T cells",
                                       "Tregs",  
                                       "Exhausted T cells", 
                                       "Cytotoxic T cells",
                                       "Natural killer cells" , 
                                       "gd T cells", 
                                       "Proliferating cells", 
                                       "other" ))

p <- SCpubr::do_DotPlot(sample = s, 
                        features = gene_list, 
                        font.size = 28, 
                        legend.length = 18,  
                        legend.type = "colorbar", 
                        group.by = "anno_l3",
                        dot.scale = 10,  
                        sequential.palette ="PiYG",
                        scale = TRUE,
                        sequential.direction = -1)
p

DT::datatable(p$data, 
              caption = ("Figure 4b"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))
```



# Figure 4c


```{r Figure 4c, fig.height=4, fig.width=8, message=FALSE, warning=FALSE,  out.width='50%'}
srat$WHO <- factor(srat$WHO, levels = c("SD","PR","CR"))


q1 <- SCpubr::do_BarPlot( s,
                         group.by = "anno_l2",
                         split.by = "WHO",
                         position = "fill",
                         flip = TRUE,
                         order=FALSE,
                         legend.position = "none",
                         font.size = 32 ,
                         colors.use = colors_anno_l2,
                         xlab = "",
                         ylab = ""
                          
                   ) + xlab("Response") + ylab("% of T and NK cells")


q1

DT::datatable(q1$data, 
              caption = ("Figure 4c"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))


```

```{r Figure 4c2, fig.height=4, fig.width=8, message=FALSE, warning=FALSE,  out.width='50%'}
q1 <- SCpubr::do_BarPlot( s,
                         group.by = "anno_l2",
                         split.by = "Pathological.Response",
                         position = "fill",
                         flip = TRUE,
                         order=FALSE,
                         legend.position = "none",
                         font.size = 32 ,
                         colors.use = colors_anno_l2,
                         xlab = "",
                         ylab = ""
                          
                   ) + xlab("Response") + ylab("% of T and NK cells")


q1

DT::datatable(q1$data, 
              caption = ("Figure 4c"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))


```

# Figure 4d

```{r Figure4d, fig.height=6, fig.width=5, message=FALSE, warning=FALSE, out.width='50%'}
s$WHO <- factor(s$WHO, levels = c("CR","PR","SD"))

df <- s@meta.data %>%
  mutate_if(sapply(s@meta.data, is.character), as.factor)  %>% 
  group_by( WHO, patient, anno_l2)%>% 
  summarise(Nb = n()) %>%
  group_by(patient) %>% 
  mutate(R1 = rep(Nb[anno_l2 == "Cytotoxic T cells"]/Nb[anno_l2 == "Tregs"],length(unique(anno_l2)))) 

df <- df[!duplicated(df$patient), c(1,2,5)]



p <- ggviolin(df, x = "WHO", y = "R1", id="patient",  color = "black", fill = "WHO" , palette= c("white","#66c2a4", "#ccece6"),add = "dotplot", point.size = 2, xlab = "WHO", ylab = " Cytotoxic T cells \n / Tregs ratio", ncol = 14) + stat_compare_means( method = "wilcox.test", paired = FALSE, label.y = c(75, 65,65), hide.ns = TRUE, label = "p.format", label.x.npc = "center", bracket.size = 0.5, comparison = list(c("CR", "SD"), c("SD", "PR"), c("CR", "PR")), size = 6)  +theme(text = element_text(size = 20)) + NoLegend() + ylim(c(0, 80))

p

DT::datatable(p$data, 
              caption = ("Figure 4d"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))

```

```{r Figure4d2, fig.height=6, fig.width=5, message=FALSE, warning=FALSE, out.width='50%'}

df <- s@meta.data %>%
  mutate_if(sapply(s@meta.data, is.character), as.factor)  %>% 
  group_by(Pathological.Response, patient, anno_l2)%>% 
  summarise(Nb = n()) %>%
  group_by(patient) %>% 
  mutate(R1 = rep(Nb[anno_l2 == "Cytotoxic T cells"]/Nb[anno_l2 == "Tregs"],length(unique(anno_l2)))) 

df <- df[!duplicated(df$patient), c(1,2,5)]



p <- ggviolin(df, x = "Pathological.Response", y = "R1", id="patient",  color = "black", fill = "Pathological.Response" , palette= c("#66c2a4", "#ccece6"),add = "dotplot", point.size = 2, xlab = "Pathological.Response", ylab = " Cytotoxic T cells \n / Tregs ratio", ncol = 14) + stat_compare_means( method = "wilcox.test", paired = FALSE, label.y = c(75), hide.ns = TRUE, label = "p.format", label.x.npc = "center", bracket.size = 0.5, comparison = list(c("pCR", "non-pCR")), size = 6)  +theme(text = element_text(size = 20)) + NoLegend() + ylim(c(0, 80))

p

DT::datatable(p$data, 
              caption = ("Figure 4d"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))

```


# Figure 4e

## Redefine `scRepertoire` function

To be able to change few ploting parameters of the `clonalHomeostasis` function, we copy/paste and modified some function from [`scRepertoire`](https://www.bioconductor.org/packages/release/bioc/html/scRepertoire.html)

```{r}
# scRepertoire functions
#Use to shuffle between chains
off.the.chain <- function(dat, chain, cloneCall) {
  chain1 <- toupper(chain) #to just make it easier
  if (chain1 %in% c("TRA", "TRD", "IGH")) {
    x <- 1
  } else if (chain1 %in% c("TRB", "TRG", "IGL")) {
    x <- 2
  } else {
    warning("It looks like ", chain, " does not match the available options for `chain = `")
  }
  dat[,cloneCall] <- str_split(dat[,cloneCall], "_", simplify = TRUE)[,x]
  return(dat)
}


#Remove list elements that contain all NA values
checkBlanks <- function(df, cloneCall) {
    for (i in seq_along(df)) {
        if (length(df[[i]][,cloneCall]) == length(which(is.na(df[[i]][,cloneCall]))) | 
            length(which(!is.na(df[[i]][,cloneCall]))) == 0 | 
            nrow(df[[i]]) == 0) {
            df[[i]] <- NULL
        } else {
            next()
        }
    }
    return(df)
}


#Ensure df is in list format
checkList <- function(df) {
    df <- if(is(df)[1] != "list") list(df) else df
    return(df)
}

checkContigs <- function(df) {
  df <- lapply(seq_len(length(df)), function(x) {
    df[[x]] <- if(is(df[[x]])[1] != "data.frame") as.data.frame(df[[x]]) else df[[x]]
    df[[x]][df[[x]] == ""] <- NA
    df[[x]] <- df[[x]][with(df[[x]], order(reads, chain)),]
  })
}

#' @importFrom dplyr bind_rows
bound.input.return <- function(df) {
  if (inherits(x=df, what ="Seurat") | inherits(x=df, what ="SummarizedExperiment")) {
    df <- grabMeta(df)
  } else {
    df <- bind_rows(df, .id = "element.names")
  }
  return(df)
}

list.input.return <- function(df, split.by) {
  if (inherits(x=df, what ="Seurat") | inherits(x=df, what ="SummarizedExperiment")) {
    if(is.null(split.by)){
      split.by <- "cluster"
    }
    df <- expression2List(df, split.by)
  } 
  return(df)
}

#Get UMAP or other coordinates
#' @importFrom SingleCellExperiment reducedDim
get.coord <- function(sc, reduction) { 
  if (is.null(reduction)) {
    reduction = "pca"
  }
  if (inherits(x=sc, what ="Seurat")) {
    coord <- sc@reductions[[reduction]]@cell.embeddings
  } else if (inherits(x=sc, what ="SummarizedExperiment")) {
    coord <- reducedDim(sc, reduction)
  }
  return(coord)
}

#This is to check the single-cell expression object
checkSingleObject <- function(sc) {
    if (!inherits(x=sc, what ="Seurat") &&
        !inherits(x=sc, what ="SummarizedExperiment")){
        stop("Object indicated is not of class 'Seurat' or 
            'SummarizedExperiment', make sure you are using
            the correct data.") }
    }

#This is to grab the meta data from a seurat or SCE object
#' @importFrom SingleCellExperiment colData 
grabMeta <- function(sc) {
    if (inherits(x=sc, what ="Seurat")) {
        meta <- data.frame(sc[[]], slot(sc, "active.ident"))
        if ("cluster" %in% colnames(meta)) {
          colnames(meta)[length(meta)] <- "cluster.active.ident"
        } else {
        colnames(meta)[length(meta)] <- "cluster"
        }
    }
    else if (inherits(x=sc, what ="SummarizedExperiment")){
        meta <- data.frame(colData(sc))
        rownames(meta) <- sc@colData@rownames
        clu <- which(colnames(meta) == "ident")
        if ("cluster" %in% colnames(meta)) {
          colnames(meta)[clu] <- "cluster.active.idents"
        } else {
          colnames(meta)[clu] <- "cluster"
        }
    }
    return(meta)
}

#This is to add the sample and ID prefixes for combineBCR()/TCR()
modifyBarcodes <- function(df, samples, ID) {
    out <- NULL
    for (x in seq_along(df)) {
        data <- df[[x]] 
        if (!is.null(ID)){
          data$barcode <- paste0(samples[x], "_", ID[x], "_", data$barcode)
        } else {
          data$barcode <- paste0(samples[x], "_", data$barcode)
        }
        out[[x]] <- data }
    return(out)
}

#Removing barcodes with NA recovered
removingNA <- function(final) {
    for(i in seq_along(final)) {
        final[[i]] <- na.omit(final[[i]])}
    return(final)
}

#Removing barcodes with > 2 clones recovered
removingMulti <- function(final){
    for(i in seq_along(final)) {
        final[[i]] <- filter(final[[i]], !grepl(";",CTnt))}
    return(final)
}

#Removing extra clonotypes in barcodes with > 2 productive contigs
#' @import dplyr
filteringMulti <- function(x) {
    x <- x %>%
      group_by(barcode, chain) %>% 
      slice_max(n = 1, order_by = reads)
    table <- subset(as.data.frame(table(x$barcode)), Freq > 2)
    if (nrow(table) > 0) {
      barcodes <- as.character(unique(table$Var1))
      multichain <- NULL
      for (j in seq_along(barcodes)) {
        chain <- x[x$barcode == barcodes[j],] %>% 
            group_by(barcode) %>% top_n(n = 2, wt = reads)
        multichain <- rbind(multichain, chain) 
        }
    x <- subset(x, barcode %!in% barcodes)
    x <- rbind(x, multichain) 
    }
    return(x)
}



#Filtering NA contigs out of single-cell expression object
#' @import dplyr
#' @importFrom SingleCellExperiment colData
filteringNA <- function(sc) {
    meta <- grabMeta(sc)
    evalNA <- data.frame(meta[,"cloneType"])
    colnames(evalNA) <- "indicator"
    evalNA <- evalNA %>%
        transmute(indicator = ifelse(is.na(indicator), 0, 1))
    rownames(evalNA) <- rownames(meta)
    if (inherits(x=sc, what ="cell_data_set")){
      colData(sc)[["evalNA"]]<-evalNA
      return(sc[, !is.na(sc$cloneType)])
    }else{
      col.name <- names(evalNA) %||% colnames(evalNA)
      sc[[col.name]] <- evalNA
      sc <- subset(sc, cloneType != 0)
      return(sc)
    }
}

#Calculating diversity using Vegan R package
#' @importFrom vegan diversity estimateR
diversityCall <- function(data) {
    w <- diversity(data[,"Freq"], index = "shannon")
    x <- diversity(data[,"Freq"], index = "invsimpson")
    y <- estimateR(data[,"Freq"])[2] #Chao
    z <- estimateR(data[,"Freq"])[4] #ACE
    z2 <- diversity(data[,"Freq"], index = "shannon")/log(length(data[,"Freq"]))
    out <- c(w,x,y,z, z2)
    return(out)
}

#Organizing list of contigs for visualization
parseContigs <- function(df, i, names, cloneCall) {
    data <- df[[i]]
    data1 <- data %>% group_by(data[,cloneCall]) %>%
        summarise(Abundance=n())
    colnames(data1)[1] <- cloneCall
    data1$values <- names[i]
    return(data1)
}

#Calculate the Morisita Index for Overlap Analysis
#' @author Massimo Andreatta, Nick Borcherding
morisitaIndex <- function(df, length, cloneCall, coef_matrix) {
    for (i in seq_along(length)){
        df.i <- df[[i]]
        df.i <- data.frame(table(df.i[,cloneCall]))
        colnames(df.i) <- c(cloneCall, 'Count')
        df.i[,2] <- as.numeric(df.i[,2])
        for (j in seq_along(length)){
            if (i >= j){ next }
            else { df.j <- df[[j]]
            df.j <- data.frame(table(df.j[,cloneCall]))
            colnames(df.j) <- c(cloneCall, 'Count')
            df.j[,2] <- as.numeric(df.j[,2])
            merged <- merge(df.i, df.j, by = cloneCall, all = TRUE)
            merged[is.na(merged)] <- 0
            X <- sum(merged[,2])
            Y <- sum(merged[,3])
            sum.df.i <- sum(df.i[,2]^2)
            sum.df.j <- sum(df.j[,2]^2)
            
            num <- 2 * sum(merged[, 2] * merged[, 3])
            den <- ((sum.df.i / (X^2) + sum.df.j / (Y^2)) * X * Y)
                
            coef.i.j <- num/den
            coef_matrix[i,j] <- coef.i.j
            }
        }
    }
    return(coef_matrix)
}


#Calculate the Jaccard Similarity Index for Overlap Analysis
jaccardIndex <- function(df, length, cloneCall, coef_matrix) {
  for (i in seq_along(length)){
    df.i <- df[[i]]
    df.i <- df.i[,c("barcode",cloneCall)]
    df.i_unique <- df.i[!duplicated(df.i[,cloneCall]),]
    for (j in seq_along(length)){
      if (i >= j){ next }
      else { 
        df.j <- df[[j]]
        df.j <- df.j[,c("barcode",cloneCall)]
        df.j_unique <- df.j[!duplicated(df.j[,cloneCall]),]
        overlap <- length(intersect(df.i_unique[,cloneCall], 
                                    df.j_unique[,cloneCall]))
        coef_matrix[i,j] <- 
          overlap/(sum(length(df.i_unique[,cloneCall]), 
                                  length(df.j_unique[,cloneCall]))-overlap)
      } 
    }
  }
  return(coef_matrix)
}

rawIndex <- function(df, length, cloneCall, coef_matrix) {
  for (i in seq_along(length)){
    df.i <- df[[i]]
    df.i <- df.i[,c("barcode",cloneCall)]
    df.i_unique <- df.i[!duplicated(df.i[,cloneCall]),]
    for (j in seq_along(length)){
      if (i >= j){ next }
      else { 
        df.j <- df[[j]]
        df.j <- df.j[,c("barcode",cloneCall)]
        df.j_unique <- df.j[!duplicated(df.j[,cloneCall]),]
        overlap <- length(intersect(df.i_unique[,cloneCall], 
                                    df.j_unique[,cloneCall]))
        coef_matrix[i,j] <- overlap
      } 
    }
  }
  return(coef_matrix)
}


#Calculate the Overlap Coefficient for Overlap Analysis
#' @author Nick Bormann, Nick Borcherding
overlapIndex <- function(df, length, cloneCall, coef_matrix) {
    for (i in seq_along(length)){
        df.i <- df[[i]]
        df.i <- df.i[,c("barcode",cloneCall)]
        df.i_unique <- df.i[!duplicated(df.i[,cloneCall]),]
        for (j in seq_along(length)){
            if (i >= j){ next }
            else { df.j <- df[[j]]
            df.j <- df.j[,c("barcode",cloneCall)]
            df.j_unique <- df.j[!duplicated(df.j[,cloneCall]),]
            overlap <- length(intersect(df.i_unique[,cloneCall], 
                                        df.j_unique[,cloneCall]))
            coef_matrix[i,j] <- 
                overlap/min(length(df.i_unique[,cloneCall]), 
                length(df.j_unique[,cloneCall])) } } }
    return(coef_matrix)
}

# This suppressing outputs for using dput()
quiet <- function(x) {
    sink(tempfile())
    on.exit(sink())
    invisible(force(x))
}

# This is to help sort the type of clonotype data to use
theCall <- function(x) {
    if (x %in% c("CTnt", "CTgene", "CTaa", "CTstrict")) {
      x <- x
    }else if (x == "gene" | x == "genes") {
        x <- "CTgene"
    } else if(x == "nt" | x == "nucleotide") {
        x <- "CTnt"
    } else if (x == "aa" | x == "amino") {
        x <- "CTaa"
    } else if (x == "gene+nt" | x == "strict") {
        x <- "CTstrict"
    }
    return(x)
}

# Assigning positions for TCR contig data
#' @author Gloria Kraus, Nick Bormann, Nick Borcherding
parseTCR <- function(Con.df, unique_df, data2) {
    for (y in seq_along(unique_df)){
        barcode.i <- Con.df$barcode[y]
        location.i <- which(barcode.i == data2$barcode)
        for (z in seq_along(location.i)) {
          where.chain <- data2[location.i[z],"chain"]
          if (where.chain == "TRA") {
            if(is.na(Con.df[y,"TCR1"])) {
              Con.df[y,tcr1_lines] <- data2[location.i[z],data1_lines]
            } else {
              Con.df[y,tcr1_lines] <- paste(Con.df[y, tcr1_lines],
                                            data2[location.i[z],data1_lines],sep=";") 
            }
          } else if (where.chain == "TRB") {
            if(is.na(Con.df[y,"TCR2"])) {
              Con.df[y,tcr2_lines] <- data2[location.i[z],data2_lines]
            } else {
              Con.df[y,tcr2_lines] <- paste(Con.df[y, tcr2_lines],
                                            data2[location.i[z],data2_lines],sep=";") 
            }
          }
        }
    }
  return(Con.df)
}

#Assigning positions for BCR contig data
#Now assumes lambda over kappa in the context of only 2 light chains
#' @author Gloria Kraus, Nick Bormann, Nick Borcherding
parseBCR <- function (Con.df, unique_df, data2) {
  for (y in seq_along(unique_df)) {
    barcode.i <- Con.df$barcode[y]
    location.i <- which(barcode.i == data2$barcode)
    if (length(location.i) == 2) {
      if (!is.na(data2[location.i[1], c("IGHct")])) {
        Con.df[y, heavy_lines] <- data2[location.i[1], h_lines]
        if(is.na(data2[location.i[2], c("IGHct")])) {
          if (!is.na(data2[location.i[2], c("IGLct")])) {
            Con.df[y, light_lines] <- data2[location.i[2], l_lines]
          } else if(!is.na(data2[location.i[2], c("IGKct")])) {
            Con.df[y, light_lines] <- data2[location.i[2], k_lines]
          }
        }
      } else if (!is.na(data2[location.i[2], c("IGHct")])) {
        Con.df[y, heavy_lines] <- data2[location.i[2], h_lines]
        if(is.na(data2[location.i[1], c("IGHct")])) {
          if (!is.na(data2[location.i[1], c("IGLct")])) {
            Con.df[y, light_lines] <- data2[location.i[1], l_lines]
          } else if(!is.na(data2[location.i[1], c("IGKct")])) {
            Con.df[y, light_lines] <- data2[location.i[1], k_lines]
          }
        }
      }
    }else if (length(location.i) == 1) {
      chain.i <- data2$chain[location.i]
      if (chain.i == "IGH") {
        Con.df[y, heavy_lines] <- data2[location.i[1], h_lines]
      }
      else if (chain.i == "IGL") {
        Con.df[y, light_lines] <- data2[location.i[1], l_lines]
      }
      else {
        Con.df[y, light_lines] <- data2[location.i[1], k_lines]
      }
    }
  }
  return(Con.df)
}

#Assign T/B cell chains and celltypes for combineTCR() and lengthContig
cellT <- function(cells) {
    if (cells == "T-AB") { 
        chain1 <- "TRA"
        chain2 <- "TRB" 
        cellType <- "T-AB" 
    } else if (cells == "T-GD") {
        chain1 <- "TRD"
        chain2 <- "TRG"
        cellType <- "T-GD" 
    } else if (cells == "B") {
        chain1 <- "IGH"
        chain2 <- "IGL"
        cellType <- "B" 
    }
    return(list(chain1, chain2, cellType))
}


#Producing a data frame to visualize for lengthContig()
lengthDF <- function(df, cloneCall, chain, group, c1, c2){
    Con.df <- NULL
    names <- names(df)
    if (chain == "both") {
            for (i in seq_along(df)) {
                length <- nchar(gsub("_", "", df[[i]][,cloneCall]))
                val <- df[[i]][,cloneCall]
                if (!is.null(group)) { 
                    cols <- df[[i]][,group]
                    data <- na.omit(data.frame(length, val, cols, names[i]))
                    colnames(data) <- c("length", "CT", group, "values")
                    Con.df<- rbind.data.frame(Con.df, data) 
                } else {
                    data <- na.omit(data.frame(length, val, names[i]))
                    colnames(data) <- c("length", "CT", "values")
                    Con.df<- rbind.data.frame(Con.df, data) }}
    } else if (chain != "both") {
            for (x in seq_along(df)) {
                df[[x]] <- off.the.chain(df[[x]], chain, cloneCall)
                strings <- df[[x]][,cloneCall]
                val1 <- strings
                for (i in seq_along(val1)) {
                    if (grepl(";", val1[i]) == TRUE) {
                        val1[i] <- str_split(val1[i], ";", simplify = TRUE)[1] 
                    } else { next() } }
                chain1 <- nchar(val1)
                if (!is.null(group)) {
                    cols1 <- df[[x]][,group]
                    data1 <- data.frame(chain1, val1, names[x], c1, cols1)
                    colnames(data1)<-c("length","CT","values","chain",group)
                }else if (is.null(group)){
                    data1 <- data.frame(chain1, val1, names[x], c1)
                    colnames(data1) <- c("length", "CT", "values", "chain")
                data <- na.omit(data1)
                data <- subset(data, CT != "NA" & CT != "")
                Con.df<- rbind.data.frame(Con.df, data) }}
    }
return(Con.df)}

#General combination of nucleotide, aa, and gene sequences for T/B cells
assignCT <- function(cellType, Con.df) {
    if (cellType %in% c("T-AB", "T-GD")) {
        Con.df$CTgene <- paste(Con.df$TCR1, Con.df$TCR2, sep="_")
        Con.df$CTnt <- paste(Con.df$cdr3_nt1, Con.df$cdr3_nt2, sep="_")
        Con.df$CTaa <- paste(Con.df$cdr3_aa1, Con.df$cdr3_aa2, sep="_")
        Con.df$CTstrict <- paste(Con.df$TCR1, Con.df$cdr3_nt1, 
            Con.df$TCR2, Con.df$cdr3_nt2, sep="_")
    } else {
        Con.df$CTgene <- paste(Con.df$IGH, Con.df$IGLC, sep="_")
        Con.df$CTnt <- paste(Con.df$cdr3_nt1, Con.df$cdr3_nt2, sep="_")
        Con.df$CTaa <- paste(Con.df$cdr3_aa1, Con.df$cdr3_aa2, sep="_") }
return(Con.df)
}


#Sorting the V/D/J/C gene sequences for T and B cells
#' @importFrom stringr str_c str_replace_na
#' @importFrom dplyr bind_rows
makeGenes <- function(cellType, data2, chain1, chain2) {
    if(cellType %in% c("T-AB", "T-GD")) {
        data2 <- data2 %>% 
            mutate(TCR1 = ifelse(chain == chain1, 
                  str_c(str_replace_na(v_gene),  str_replace_na(j_gene), str_replace_na(c_gene), sep = "."), NA)) %>%
            mutate(TCR2 = ifelse(chain == chain2, 
                  str_c(str_replace_na(v_gene), str_replace_na(d_gene),  str_replace_na(j_gene),  str_replace_na(c_gene), sep = "."), NA))
    }
    else {
        heavy <- data2[data2$chain == "IGH",] %>% 
          mutate(IGHct = str_c(str_replace_na(v_gene), str_replace_na(d_gene),  str_replace_na(j_gene),  str_replace_na(c_gene), sep = "."))
        kappa <- data2[data2$chain == "IGK",] %>% 
          mutate(IGKct = str_c(str_replace_na(v_gene),  str_replace_na(j_gene), str_replace_na(c_gene), sep = ".")) 
        lambda <- data2[data2$chain == "IGL",] %>%
          mutate(IGLct = str_c(str_replace_na(v_gene),  str_replace_na(j_gene), str_replace_na(c_gene), sep = "."))
        data2 <- bind_rows(heavy, kappa, lambda)
    }
    return(data2)
}

short.check <- function(df, cloneCall) {
  min <- c()
  for (x in seq_along(df)) {
    min.tmp <- length(which(!is.na(unique(df[[x]][,cloneCall]))))
    min <- c(min.tmp, min)
  }
  min <- min(min)
  return(min)
}

select.gene <- function(df, chain, gene, label) {
  if (chain %in% c("TRB", "TRG", "IGH")) {
    gene <- unname(c(V = 1, D = 2, J = 3, C = 4)[gene])
  } else {
    gene <- unname(c(V = 1, J = 2, C = 3)[gene])
  }
  if (ncol(str_split(df[,"CTgene"], "_", simplify = TRUE)) == 1) {
    C1 <- str_split(df[,"CTgene"], "_", simplify = TRUE)[,1] 
    C1 <- str_split(C1, "[.]", simplify = TRUE)[,gene] 
    df$C1 <- C1
    x <- "C1"
  } else {
    C1 <- str_split(df[,"CTgene"], "_", simplify = TRUE)[,1] 
    C1 <- str_split(C1, "[.]", simplify = TRUE)[,gene] 
    C2 <- str_split(df[,"CTgene"], "_", simplify = TRUE)[,2] 
    C2 <- str_split(C2, "[.]", simplify = TRUE)[,gene] 
    df$C1 <- C1
    df$C2 <- C2
    if (chain %in% c("TRA", "TRD", "IGH")) {
      x <- "C1"}
    else if (chain %in% c("TRB", "TRG", "IGL")) {
      x <- "C2"}
  }
  return(df)
}



#' Examining the clonal homeostasis
#'
#' This function calculates the space occupied by clonotype proportions. 
#' The grouping of these clonotypes is based on the parameter cloneTypes, 
#' at default, cloneTypes will group the clonotypes into bins of Rare = 0 
#' to 0.00005, Small = 0.00005 to 0.0005, etc. To adjust the proportions, 
#' change the number or labeling of the cloneTypes paramter. If a matrix 
#' output for the data is preferred, set exportTable = TRUE.
#'
#' @examples
#' #Making combined contig data
#' x <- contig_list
#' combined <- combineTCR(x, rep(c("PX", "PY", "PZ"), each=2), 
#' rep(c("P", "T"), 3), cells ="T-AB")
#' clonalHomeostasis(combined, cloneCall = "gene")
#'
#' @param df The product of combineTCR(), combineBCR(), expression2List(), or combineExpression().
#' @param cloneTypes The cutpoints of the proportions.
#' @param cloneCall How to call the clonotype - VDJC gene (gene), 
#' CDR3 nucleotide (nt), CDR3 amino acid (aa), or 
#' VDJC gene + CDR3 nucleotide (strict).
#' @param chain indicate if both or a specific chain should be used - 
#' e.g. "both", "TRA", "TRG", "IGH", "IGL"
#' @param exportTable Exports a table of the data into the global 
#' environment in addition to the visualization
#' @import ggplot2
#' @importFrom stringr str_split
#' @importFrom reshape2 melt
#' @export
#' @return ggplot of the space occupied by the specific proportion of clonotypes



clonalHomeostasis_m <- function(df, cloneTypes = c(Rare = .00005, Small = .0005, 
                        Medium = .005, Large = .05, Hyperexpanded = 1),
                        cloneCall = "strict", chain = "both", 
                        exportTable = FALSE) {
    cloneTypes <- c(cloneTypes)
    df <- list.input.return(df)
    cloneCall <- theCall(cloneCall)
    df <- checkList(df)
    df <- checkBlanks(df, cloneCall)
    mat <- matrix(0, length(df), length(cloneTypes) - 1, 
                dimnames = list(names(df), 
                names(cloneTypes)[-1]))
    if (chain != "both") {
      for (x in seq_along(df)) {
        df[[x]] <- off.the.chain(df[[x]], chain, cloneCall)
      }
    }
    df <- lapply(df, '[[', cloneCall)
    df <- lapply(df, na.omit)
    fun <- function(x) { table(x)/length(x) }
    df <- lapply(df, fun)
    for (i in 2:length(cloneTypes)) {
        mat[,i-1] <- vapply(df, function (x) sum(x[x > cloneTypes[i-1] & x <= 
                            cloneTypes[i]]), FUN.VALUE = numeric(1))
        colnames(mat)[i-1] <- paste0(names(cloneTypes[i]), ' (', 
                                    cloneTypes[i-1], ' < X <= ', 
                                    cloneTypes[i], ')') }
    if (exportTable == TRUE) { return(mat) }
    mat_melt <- melt(mat)
    col <- length(unique(mat_melt$Var2))
   order_mat <- mat_melt %>%
    filter(Var2 == "Hyperexpanded (0.05 < X <= 1)") %>%
     arrange(value, desc = "FALSE")
   o <- order_mat$Var1
    
    
    mat_melt <- mat_melt[mat_melt$Var1 %in% o,]    
    mat_melt$Var1 <- factor(mat_melt$Var1, levels = o)
    
    plot <- ggplot(mat_melt, aes(x=as.factor(Var1), y=value, fill=Var2)) +
        geom_bar(stat = "identity", position="fill", 
                    color = "black", lwd= 0.25) +
        scale_fill_manual(name = "Clonotype Group", 
                    values = c(  "#8c6bb1","#41b6c4", "#fec44f","#ce1256" )) +
        xlab("Samples") +
        ylab("Relative Abundance") +
        theme_void() +  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 20),axis.text.y = element_text(size = 20), legend.text = element_text(size = 20), axis.title = element_text(size=20), axis.title.y = element_text(angle = 90))

    return(plot)
}

```

## Open clonotype data

n= 29690 TCR with EXACTLY one alpha chain CDR3 and one beta chain CDR3 

```{r clono, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
if(params$accessibility == "unlock"){

# loading the data
  vdj <- list()
  dirs <- list.dirs(path = path_to_data, full.names = TRUE, recursive = FALSE)

  samples <- c("NeoBCC007_post",
             "NeoBCC008_post",
             "NeoBCC012_post",
             "NeoBCC017_post",
             "NeoBCC004_post",
             "NeoBCC005_post",
             "NeoBCC006_post",
             "NeoBCC010_post",
             "NeoBCC011_post",
             "NeoBCC014_post",
             "NeoBCC015_post",
             "NeoBCC018_post")


  for (i in (samples)){
  dir_TCR <- paste0(path_to_data, i, "_abTCR_VDJ")
  vdj[[i]] <- read.csv(paste0(dir_TCR, "/filtered_contig_annotations.csv"))
  }

  head(vdj[[1]])

# combine the contigs
  combined <- combineTCR(vdj, 
                samples = gsub("_pre|_post", "", samples), 
                cells = "T-AB", removeNA = TRUE, removeMulti = TRUE)

 merge_combined <- bind_rows(combined, .id = "sample")

 saveRDS(combined, "../metadata/100_CombinedTCR.Rds")
} else{
  print("Please request access to the BCR-Seq data.")
}
```

## Add transcriptomic data
```{r echo=TRUE, fig.height=10, fig.width=20, message=FALSE, warning=FALSE, out.width='100%'}

if(params$accessibility == "unlock"){

  srat <- RenameCells(srat,new.names = paste0(srat$patient, "_",gsub("(_1|_2|_3|_4|_5|_6|_7|_8|_9)*", "",colnames(srat))))
  srat <- RenameCells(srat,new.names = gsub("_post", "",colnames(srat)))

  srat$sample <- gsub("_post", "",srat$patient)

  seurat <- combineExpression(combined, srat, 
                  cloneTypes = c(Rare = .00005, Small = .0005, 
                        Medium = .005, Large = .05, Hyperexpanded = 1),
                        cloneCall = "strict", 
                  group.by = "sample", chain = "both",
                  proportion = TRUE,
                  filterNA = TRUE)



  s    <- subset(seurat, subset = anno_l1 %in% c("CD8+ T cells","CD4+ T cells", "Proliferating cells", "Tregs", "Natural killer cells"))

  s$cloneType <- droplevels(s$cloneType)

  s@meta.data$Pathological.Response <- factor(s@meta.data$Pathological.Response, levels = c("pCR", "non-pCR"))



} else{
  print("Please request access to the BCR-Seq data.")
}
```

We mapped  26528 TCR to the scRNA-Seq data, including 23099 to annotated T cells (10699 CD4 T cells, 1951 Tregs and 9517 CD8 T cells).
Only 3 patients do not have hyperexpanded clones (2 non-pCR and 1 pCR). 



```{r echo=TRUE, fig.height=7, fig.width=25, message=FALSE, warning=FALSE, out.width='100%'}
if(params$accessibility == "unlock"){

  ss <- subset(seurat, subset = anno_l1 %in% c("CD8+ T cells") )
  ss@meta.data$cloneType <- factor(ss@meta.data$cloneType, levels = c("Hyperexpanded (0.05 < X <= 1)", "Large (0.005 < X <= 0.05)", "Medium (5e-04 < X <= 0.005)", "Small (5e-05 < X <= 5e-04)"))
  # define cell group membership
  Idents(ss) <- ss$cloneType

  de_markers <- DElegate::FindAllMarkers2(ss, replicate_column = "patient", method = "edger", min_fc = log2(1), min_rate = 0.1)



  signature_h <- de_markers$feature[de_markers$group1=="Hyperexpanded (0.05 < X <= 1)"]
  background <- rownames(ss)

  # enricher for cnet plot
  GO_H_gene_sets = msigdbr(species = "human", category = "H")
  msigdbr_t2g = GO_H_gene_sets %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()

  ego_h <- enricher(gene = signature_h, universe = background, TERM2GENE = msigdbr_t2g)

 
  b <- barplot(ego_h, title = "HallmarksÂ´enrichment of hyperexpanded clones")


  tmp <- b$data
  
  tmp$ID <- str_trunc(tmp$ID, 100, "right")
  tmp$ID <- factor(tmp$ID, levels = tmp$ID[order(tmp$Count)])

  p <- ggplot(tmp, aes(Count, ID)) +
    geom_bar(stat = "identity", color="black", fill =  colors_clonotype["Hyperexpanded (0.05 < X <= 1)"]) +
   theme_classic()+ 
    geom_text(
        aes(label = (paste( "padj=", formatC(p.adjust, format = "e", digits = 2)))),
        color = "black",
        size = 6,
        hjust=1,
        position = position_dodge(0.5)  ) +
     theme(text = element_text(size=24)) +
  ggtitle(" Hyperexpanded \n CD8+ T cells clones")

  print(p)


DT::datatable(p$data, 
              caption = ("Figure 4e"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))
} else{
  print("Please request access to the BCR-Seq data.")
}
```


# Figure 4f

We maped 1286 TRA out of 23099 detected TRA (7.3%)
We maped 435 unique TRA out of 10741 unique TRA chains (4%). 

```{r echo=TRUE, fig.height=10, fig.width=20, message=FALSE, warning=FALSE, out.width='50'}
if(params$accessibility == "unlock"){

  McPAS <- utils::read.csv(params$McPAS)
  head(McPAS)

  a <- colsplit(s$CTaa, "_", name=c("CTaa_TRA", "CTaa_TRB"))
  rownames(a) <- colnames(s)
  a$rid <- rownames(a)


  c <- McPAS[McPAS$CDR3.alpha.aa %in%  na.omit(a$CTaa_TRA), c("CDR3.alpha.aa", "Category", "Pathology")]
  head(c)
  colnames(c) <- c("CTaa_TRA",  "Category", "Pathology")



  joint <- left_join(a, c, by = "CTaa_TRA")
  joint <- joint[joint$rid %in% colnames(s),]
  joint <- joint[!duplicated(joint$rid),]
  rownames(joint) <- (joint$rid)
  
  s <- AddMetaData(s, joint)

  Idents(s) <- "Pathology"
  cells <- WhichCells(s, ident =  c("Human herpes virus 1" ) )

  Idents(s) <- "Category"
  s <- SetIdent(s, cells = cells, value = paste0("hHSV: ", s@meta.data$CTaa[s@meta.data$Pathology %in% c("Human herpes virus 1")]))
  s$Category <- Idents(s)

  Idents(s) <- "Pathology"
  cells <- WhichCells(s, ident =  c("Herpes simplex virus 2 (HSV2)") )
  Idents(s) <- "Category"
  s <- SetIdent(s, cells = cells,  value = paste0("hHSV: ", s@meta.data$CTaa[s@meta.data$Pathology %in% c("Herpes simplex virus 2 (HSV2)")]))
  s$Category <- Idents(s)

} else{
  print("Please request access to the BCR-Seq data.")
}

```





```{r Figure4f, echo=TRUE, fig.height=10, fig.width=15, message=FALSE, warning=FALSE, out.width='50%'}
if(params$accessibility == "unlock"){

  Idents(s) <- "Category"
  cells <- WhichCells(s, ident =  c("Pathogens","Autoimmune", "Allergy" ) )
  Idents(s) <- "Category"
  s <- SetIdent(s, cells = cells, value = "Other")
  s$Category <- Idents(s)
  s <- subset(s, subset = Category != "Other")
  s$Category <- droplevels(s$Category)

  p3 <- SCpubr::do_BarPlot(subset(s, subset = Category != "NA" & Category != "Other"  & Pathological.Response == "pCR"),
                         group.by = "Category",
                         split.by = "patient",
                         order=FALSE,
                         legend.position = "right",
                        colors.use = colors_McPAS,
                        font.size = 28, flip = TRUE) + xlab("pCR")

  p4 <- SCpubr::do_BarPlot(subset(s, subset = Category != "NA" & Category != "Other"& Pathological.Response == "non-pCR"),
                         group.by = "Category",
                         split.by = "patient",
                         order=FALSE,
                         legend.position = "right",
                         colors.use = colors_McPAS,
                   font.size = 28, flip = TRUE) + xlab("non-pCR")

  p3$labels$y <- " "
  
  p <- p3 + p4 + plot_layout(ncol = 1,  heights   = c(1,2),  guides = "collect") + ylab("Number of mapped TCR")
  print(p)

 
  DT::datatable(rbind(p3$data, p4$data), 
              caption = ("Figure 4f"),
              extensions = 'Buttons', 
              options = list( dom = 'Bfrtip',
              buttons = c( 'csv', 'excel')))

} else{
  print("Please request access to the BCR-Seq data.")
}

```






# Session Info

```{r , fig.height=8, fig.width=20, message=FALSE, warning=FALSE,  out.width='100%'}
sessionInfo()
```

